### PR TITLE
CustomSelectControlV2: Use `InputBase` for styling

### PR DESCRIPTION
**Pull request by mirka**
After #60226
Part of #55023

## What?

Reuse `InputBase` and `SelectControlChevronDown` for styling `CustomSelectControlV2`.

## Why?

For more DRY and consistent styling.

(`InputBase` is an internal component used to style the standard borders for an input, as well as handle the layout for prefix/suffix elements.)

## Testing Instructions

- Focus styles for CustomSelectControlV2 Legacy should match the styles for v1.
- The chevron should have the correct padding in all possible sizes.

> [!NOTE]  
> Long strings will still overlap with the chevron. I'll address this separately.

### CODE REVIEW

File: packages/components/src/custom-select-control-v2/custom-select.tsx
Comment (mirka) [inline packages/components/src/custom-select-control-v2/custom-select.tsx:92]: We don't need to connect this to the context system anymore.

File: packages/components/src/custom-select-control-v2/legacy-component/index.tsx
Comment (mirka) [inline packages/components/src/custom-select-control-v2/legacy-component/index.tsx:124]: I replaced this with direct prop assignments on the `_CustomSelect` component for simplicity.

File: packages/components/src/custom-select-control-v2/legacy-component/index.tsx
Comment (mirka) [inline packages/components/src/custom-select-control-v2/legacy-component/index.tsx:118]: `describedBy` wasn't being destructured from the props, and therefore it was still being passed invalidly to `_CustomSelect` through the rest props. Fixed.

File: packages/components/src/custom-select-control-v2/legacy-component/index.tsx
Comment (mirka) [inline packages/components/src/custom-select-control-v2/legacy-component/index.tsx:113]: No more need to tunnel this via context system, because we can directly set the newly exposed `size` prop on `_CustomSelect`.

File: packages/components/src/input-control/input-base.tsx
Comment (mirka) [inline packages/components/src/input-control/input-base.tsx:67]: Removed this unused export, because a consumer can erroneously import this named version instead of the default export which is properly wrapped in a `forwardRef()`.

(This happened to me while working on this PR and it wasted a couple minutes of my life 😇)

File: packages/components/src/custom-select-control-v2/types.ts
Comment (mirka) [inline packages/components/src/custom-select-control-v2/types.ts:1]: I refactored some of these types because the rejiggering work I did in this PR surfaced some mistakes in how we were using them.

File: packages/components/src/custom-select-control-v2/types.ts
Comment (mirka) [inline packages/components/src/custom-select-control-v2/types.ts:17]: For context, the public interface of `CustomSelectControlV2` only has these two size variants, whereas the private (internal) `CustomSelectButton` has three size variants for back compat.

File: packages/components/src/custom-select-control-v2/custom-select.tsx
Comment (fullofcaffeine) [reply packages/components/src/custom-select-control-v2/custom-select.tsx:92]: Would you mind elaborating on why?

File: packages/components/src/custom-select-control-v2/legacy-component/index.tsx
Comment (fullofcaffeine) [reply packages/components/src/custom-select-control-v2/legacy-component/index.tsx:113]: Ah I see 👍🏻 (regarding https://github.com/WordPress/gutenberg/pull/60261/files#r1616307373).

File: packages/components/src/input-control/input-base.tsx
Comment (fullofcaffeine) [reply packages/components/src/input-control/input-base.tsx:67]: Nit, but, what is the current convention for components - should we provide both (named, default) or just one - and if so, it seems `default` is preferred?

Another approach would be to connect before while assigning to a `const` and then export as needed.

**Review by fullofcaffeine — APPROVED** [review]
I didn't know about `InputBase`; I can see it's part of the `input-control` component. Would it make more sense to make it a separate component if it's going to be used as the basis for others like this (in a follow-up, ofc)? It feels weird it doesn't have its own doc, too. For now, it'd be nice if you could add a bit more information/background on `InputBase` in the PR description, if possible. 

Other than that, the changes looks good! Storybook stories also look good. However, I can still see the API and styling issues I've found while working on https://github.com/WordPress/gutenberg/pull/61272. Since this component is still private, I think it's fine to merge this now and rebase/merge it into https://github.com/WordPress/gutenberg/pull/61272 to see how it behaves with the fixes there.  Will start looking into it as soon as this is merged.

File: packages/components/src/custom-select-control-v2/custom-select.tsx
Comment (fullofcaffeine) [reply packages/components/src/custom-select-control-v2/custom-select.tsx:92]: Nevermind, already seen https://github.com/WordPress/gutenberg/pull/60261#discussion_r1542085948 👍🏻

File: packages/components/src/custom-select-control-v2/legacy-component/index.tsx
Comment (tyxla) [inline packages/components/src/custom-select-control-v2/legacy-component/index.tsx:94]: Is there a good reason for the IIFE here? Can be a simple `if` without an IIFE IMHO.

File: packages/components/src/custom-select-control-v2/legacy-component/index.tsx
Comment (tyxla) [inline packages/components/src/custom-select-control-v2/legacy-component/index.tsx:117]: Do we intentionally allow the `restProps` to override the above props like `store` or `renderSelectedValue`? Should these props be declared below the `restProps` to prevent that from happening?

**Review by tyxla — APPROVED** [review]
Looks good to me visually and works well 👍 Thanks.

Just left a couple of minor questions, but other than that, good to go 🚀

File: packages/components/src/input-control/input-base.tsx
Comment (mirka) [reply packages/components/src/input-control/input-base.tsx:67]: Good question. We do indeed have [some conventions](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/CONTRIBUTING.md#considerations-for-the-docgen) that are actually for accommodating our current docgen, more than anything else.

Doesn't matter that much for internal components like `InputBase`, which aren't fed into the docgen.

File: packages/components/src/custom-select-control-v2/legacy-component/index.tsx
Comment (DaniGuardiola) [reply packages/components/src/custom-select-control-v2/legacy-component/index.tsx:117]: Everything that can't be merged should be overridable, as per the open/closed pattern.

File: packages/components/src/custom-select-control-v2/legacy-component/index.tsx
Comment (mirka) [reply packages/components/src/custom-select-control-v2/legacy-component/index.tsx:94]: To me, this takes the least cognitive load to read because it doesn't introduce a mutable variable (`let`), and it's a bit of overkill to extract out the entire function. I'm probably more mutation-averse than average though 😓

File: packages/components/src/custom-select-control-v2/legacy-component/index.tsx
Comment (mirka) [reply packages/components/src/custom-select-control-v2/legacy-component/index.tsx:117]: Also `store` or `renderSelectedValue` are not documented in the public interface for this, so it would be very at-your-own-risk if anyone tried that.

No strong opinion, but I think the general strategy of "keep everything overridable unless there is a specific, strong reason not to" would be fine.

File: packages/components/src/custom-select-control-v2/legacy-component/index.tsx
Comment (tyxla) [reply packages/components/src/custom-select-control-v2/legacy-component/index.tsx:94]: Primitive values cannot be mutated, so we're technically not really mutating anything. Also, the IIFE makes the code a bit more complex to read than if we were using a `let`. No strong feelings either way, so I'm happy to leave this to your judgment!

Comment (mirka) [reply packages/components/src/custom-select-control-v2/legacy-component/index.tsx:94]: Touché, I should've said _reassignment_! I always considered variable reassignment to be a form of mutation (as in mutable state), but I guess those two terms should be differentiated in JS at least.

**Comment by mirka** [comment]
> I didn't know about `InputBase`; I can see it's part of the `input-control` component. Would it make more sense to make it a separate component if it's going to be used as the basis for others like this (in a follow-up, ofc)? It feels weird it doesn't have its own doc, too. For now, it'd be nice if you could add a bit more information/background on `InputBase` in the PR description, if possible.

Sorry about that. I added a quick note to the PR description, and a JSDoc on the component for good measure. And good point, we might want to consider extracting out the `InputBase` from the `input-control` folder, since it's starting to be used in more places.

**Comment by DaniGuardiola** [comment]
@mirka fwiw, I wanna take a closer look at this InputBase thingy. A while back I looked at it and it seems to be doing a whole lot of not really that much :P - I think there's potential for simplifying or splitting this up into something more manageable and way simpler. Note that I'm including a few different layers that I noticed in input code, so maybe I'm getting things mixed up here and this one's as simple as it gets. In any case, this is for another day, just wanted to make a note!

File: packages/components/src/custom-select-control-v2/legacy-component/index.tsx
Comment (fullofcaffeine) [reply packages/components/src/custom-select-control-v2/legacy-component/index.tsx:117]: Would be much better if JS supported `if` expressions :/

**Pull request merged**
